### ISSUE TITLE
Running external build systems

### ISSUE THREAD

**Issue by ramblehead** [aspect-build/bazel-lib#228]
Coming here from this [SO](https://stackoverflow.com/a/72661140) answer.

I am building a template/POC repository for my future C++-centred projects:
https://github.com/rh-ogorod/diamond-dojo/tree/main

The current attempt is to use Bazel for building/testing/deployment coordination - locally and at CI services. My main requirements are: automatic ```compile_commands.json``` generation - usable with clang-family dev tools; and ruining external build systems for dependency projects when those projects are not supporting Bazel officially.

```compile_commands.json``` solution has been prototyped with customised Bazel rule from GRAIL and my unbox.js script ([bazelbuild-rules-compdb](https://github.com/rh-ogorod/bazelbuild-rules-compdb.git) - it is used in the above diamond-dojo project).

Running external build systems with Bazel is difficult. So far I found three possible strategies (excluding writing custom rules):
1. Building external deps before running Bazel and then capturing external build artefacts into Bazel via ```cc_library```. This strategy has been the most practical.
2. Using ```rules_foreign_cc```.
3. Using external build ```genrule``` as a label in ```data``` attribute of ```cc_library``` target.

"pre-bazel" building strategy is not perfect as it is not hermetic (depending on external build system) and somewhat defeats the usage of Bazel.
```rules_foreign_cc``` strategy although covers many practical cases, is not sufficiently generic - even some ```cmake``` and ```make``` projects are not usable (e.g. with internal custom scripts or unconventional dependencies management). Also, when Bazel-managed external tool uses multi-threading with its own memory management, the host system may easily run out of resources.
```genrule``` strategy requires listing of all its artefact files in corresponding ```cc_library``` ```hdrs``` and ```srcs``` attributes that is time-consuming and not practical when experimenting/playing with new libraries.

In the above diamond-dojo repository, boost's b2 is used as an external build system model (I am aware of nelhage Bazel boost rules). Is it possible to use aspect's ```run_binary``` to replace ```genrule``` when running arbitrary build systems without the need to list all artefact files in corresponding ```cc_library``` ```hdrs``` and ```srcs``` attributes? Or "somehow" couple ```run_binary``` output directly with ```cc_library```/```cc_binary```? Something like:

https://github.com/rh-ogorod/diamond-dojo/blob/340f09f5b36a7ea04e865b5d19bc08b86b1eef46/external/boost/BUILD.bazel#L196 (not working!)

**Comment by ramblehead** [comment]
After a few more experiments, the following ```run_binary``` solution seems to be working:

```starlark
run_binary(
  name = "build",
  mnemonic = "BuildBoost",
  srcs = [":all_srcs"],
  outs = [
    "dist/lib/libboost_system-gcc11-mt-x64-1_80.a",
    "dist/lib/libboost_system-gcc11-mt-x64-1_80.so.1.80.0",
    "dist/lib/libboost_system-gcc11-mt-x64-1_80.so",
  ],
  out_dirs = ["package", "dist/include", "build"],
  tool = ".rh-subproject/bazel-build.sh",
  env = {"RULEDIR": "$(RULEDIR)"},
  execution_requirements = { "exclusive": "1" }, # not sure if it has any effect here
  visibility = ["//visibility:public"],
)

cc_library(
  name = "boost",
  hdrs = glob(["package/boost/**"]),
  strip_include_prefix = "package/",
  srcs = [
    "dist/lib/libboost_system-gcc11-mt-x64-1_80.a",
    "dist/lib/libboost_system-gcc11-mt-x64-1_80.so.1.80.0",
    "dist/lib/libboost_system-gcc11-mt-x64-1_80.so",
  ],
  data = [":build"],
  visibility = ["//visibility:public"],
)
```
(see https://github.com/rh-ogorod/diamond-dojo/blob/62009bfebb10896b9819d792d0e7b2014cd7ac66/external/boost/BUILD.bazel)

The above is a bit better than what I was getting with ```genrule``` - ```cc_library``` is now glob'ing headers from its data "build" target. Libraries (.a  and .so), however, still should be listed manually.

The above result brought me to an idea of creating a second ```run_binary``` target that can depend on the "build" target and which outputs may be fed directly into ```cc_library``` ```srcs```, like the following:

```starlark
run_binary(
  name = "build",
  mnemonic = "BoostBuild",
  srcs = [":all_srcs"],
  out_dirs = ["package", "dist", "build"],
  tool = ".rh-subproject/bazel-build.sh",
  env = {"RULEDIR": "$(RULEDIR)"},
  execution_requirements = { "exclusive": "1" }
  visibility = ["//visibility:public"],
)

run_binary(
  name = "build_copy_lib",
  mnemonic = "BoostBuildCopyLibs",
  srcs = [":build"],
  # outs = ["lib_copy/libboost_system-gcc11-mt-x64-1_80.a"],
  out_dirs = ["lib_copy"],
  tool = ".rh-subproject/bazel-build-copy-lib.sh",
  env = {"RULEDIR": "$(RULEDIR)"},
  visibility = ["//visibility:public"],
)

cc_library(
  name = "boost",
  hdrs = glob(["package/boost/**"]),
  strip_include_prefix = "package/",
  srcs = [
    ":build_copy_lib",
  ],
  data = [":build"],
  visibility = ["//visibility:public"],
)
```
(see https://github.com/rh-ogorod/diamond-dojo/blob/39de55f2301769f9b1f04b521a84b7356a8acbad/external/boost/BUILD.bazel)

Unfortunately, above ```build_copy_lib``` approach produces the following error from bazel:

`
ERROR: Unexpected exception, please file an issue with the Bazel team: Artifact 'bazel-out/k8-fastbuild/bin/external/boost/lib_copy/libboost_system-gcc11-mt-x64-1_80.a' expanded from the directory artifact 'bazel-out/k8-fastbuild/bin/external/boost/lib_copy' is neither header nor source file.
`

If in ```build_copy_lib```, ```out_dirs``` is replaced with concrete ```outs``` (e. g. ```["lib_copy/libboost_system-gcc11-mt-x64-1_80.a"]```), everything works as expected...

**Comment by kormide** [comment]
@ramblehead If you look in the output tree at `bazel-out/k8-fastbuild/bin/external/boost/lib_copy/libboost_system-gcc11-mt-x64-1_80.a` does anything seem off about it? Is it the same file when you use `outs` instead of `out_dirs`?

**Comment by ramblehead** [comment]
@kormide ```bazel-out/k8-fastbuild/bin/external/boost/lib_copy/libboost_system-gcc11-mt-x64-1_80.a``` looks normal.

I have created another simplified repository, eliminating most of auxiliary bashes in favour of .bazelrc and platforms: 
https://github.com/rh-ogorod/coding-dojo

In ```main``` branch of the above repo boost is built non-hermetically before running bazel build (this is how it is intended to work):
```bash
# building
git clone git@github.com:rh-ogorod/coding-dojo.git  # current commit - 1088fc0
cd coding-dojo
./bootstrap.sh

# usage example
bazel run //packages/binary-gap/cpp20:main
```

In ```boost``` branch of the coding-dojo repo attempt is made to build boost hermetically using ```run_binary```:
```bash
# from coding-dojo dir:
bazel clean --expunge
git checkout boost  # current commit - fb90324
git submodule update --init --recursive
./external/boost/.rh-subproject/clean.sh
bazel build @boost//:boost --verbose_failures --sandbox_debug
```

Running above script produces ```ERROR: Unexpected exception, please file an issue with the Bazel team: Artifact 'bazel-out/k8-fastbuild/bin/external/boost/lib_copy/libboost_prg_exec_monitor-gcc11-mt-x64-1_80.a' expanded from the directory artifact 'bazel-out/k8-fastbuild/bin/external/boost/lib_copy' is neither header nor source file.```

Changing ```out_dirs``` to ```outs``` in boost ```run_binary``` builds with no errors. However, running the following:
```bash
bazel test //packages/binary-gap/cpp20:main-test
```

results in the following compilation error:
```
packages/binary-gap/cpp20/src/main.test.cpp:6:10: fatal error: boost/test/unit_test.hpp: No such file or directory
    6 | #include <boost/test/unit_test.hpp>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

Meaning that my original statement that ```glob``` works for headers was wrong - it just does not produce error during build. Attempts to move headers to ```out_dirs``` results in the same "is neither header nor source file" error.

**Comment by alexeagle** [comment]
This `neither header nor source file` error comes from some other ruleset which appears to not understand directory inputs.

Do you have a reason to think there's a bug in bazel-lib?

**Comment by ramblehead** [comment]
> Do you have a reason to think there's a bug in bazel-lib?

This is what I am trying to understand. The initial idea was to find “a simple” way to avoid explicitly specifying files in stock bazel rules – such as ```cc_library```’s ```hdrs``` and ```srcs``` when those files are coming as outputs from other toolchains “outside of bazel”. That is how I came across the original SO about ```run_binary```.

My hope was to use ```out_dirs``` instead of ```outs``` so ```run_binary``` would “somehow” glob ```out_dirs``` and provide files contained in rule’s output to the dependant rules as if those files were explicitly listed in ```outs```.

I did not attempt to go through ```run_binary``` rule source code. Because dependant ```cc_library``` complains about files inside ```out_dirs``` (e.g. ```libboost_system-gcc11-mt-x64-1_80.a```) - not the ```out_dirs``` directories themselves, it seemed that ```run_binary``` was forwarding files from ```out_dirs```.

@alexeagle from your reply it looks that ```run_binary``` is not intended to provide files containing in ```out_dirs``` instead of explicitly listing those files in ```outs``` (i.e. dependant rules should be explicitly designed to work with ```out_dirs```). Is it correct?
